File: 127-word-ladder/word-ladder.py
class Solution:
    def compare_words(self, word1:str, word2:str):
        diff_count = 0
        i = 0
        n = len(word1)
        for i in range(n):
            if word1[i] != word2[i]:
                diff_count += 1
        return diff_count

    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordList.append(beginWord)
        if endWord not in wordList:
            return 0
        adj = {word:[] for word in wordList}
        n = len(wordList)

        for i in range(n):
            for j in range(i+1, n):
                if self.compare_words(wordList[i], wordList[j]) == 1:
                    adj[wordList[i]].append(wordList[j])
                    adj[wordList[j]].append(wordList[i])
        
        queue = deque()
        visited = set()
        queue.append(beginWord)
        visited.add(beginWord)
        distance = 1
        distances = {word:0 for word in wordList}

        while(queue):
            i = len(queue)
            while(i):
                node = queue.popleft()
                for child in adj[node]:
                    if child not in visited:
                        distances[child] = distance
                        visited.add(child)
                        queue.append(child)


                i-=1
            distance += 1
        return distances[endWord]+1 if distances[endWord] != 0 else 0

File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    bool isOneDiff(string word1, string word2) {
        int n = word1.size();
        if(n != word2.size()) return false;
        
        int diff = 0;

        for(int i = 0;i < n && diff <=1; i++){
            if(word1[i] != word2[i]) diff += 1;
        }

        return diff == 1?true:false;
    }

    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        int n = wordList.size();
        bool end_found = false;
        unordered_map<string, vector<string>> graph;
        for(int i = 0;i < n; i++){
            if(wordList[i] == endWord) end_found = true;
            for(int j = i+1; j < n; j++){
                if(isOneDiff(wordList[i],wordList[j])){
                    graph[wordList[i]].push_back(wordList[j]);
                    graph[wordList[j]].push_back(wordList[i]);
                }
            }
        }
        for (int j = 0; j < n; j++)
        {
            if (isOneDiff(beginWord, wordList[j]))
            {
                graph[beginWord].push_back(wordList[j]);
                graph[wordList[j]].push_back(beginWord);
            }
        }

        if (!end_found)
            return 0;

        queue<string> q({beginWord});
        unordered_set<string> visited;
        visited.insert(beginWord);
        int ans = 0;
        while (!q.empty())
        {
            ans += 1;
            int cur_queue_size = q.size();
            while (cur_queue_size--)
            {
                string cur_word = q.front();
                q.pop();
                for (const auto &child_word : graph[cur_word])
                {
                    if (child_word == endWord)
                        return ans+1;
                    if (!visited.contains(child_word))
                    {
                        visited.insert(child_word);
                        q.push(child_word);
                    }
                }
            }
        }

        return 0;
    }
};